### PR TITLE
Fix #44 - Problems using FSI on a project

### LINKED ISSUE #44: Problems using FSI on a project
Opened at CodePlex by ovatsus
[from https://github.com/fsharp/fsharp/issues/278](from https://github.com/fsharp/fsharp/issues/278) 

Let's say we have an F# project consistent of two files (this is the minimum to be able reproduce this): 

File A.fs:
```fsharp
namespace Project

type DU = A | B
```
File B.fs:
```fsharp
namespace Project

type B = { Prop : DU }
```
This compiles fine. If you know want to test this on a script file:
```fsharp
#load "A.fs"
#load "B.fs"
```
If you evaluate this, you'l get this error:
[Loading c:\users\guguer\documents\visual studio 2013\Projects\ConsoleApplication2\A.fs]
```fsharp
namespace FSI_0002.Project
  type DU =
    | A
    | B
``
> 
[Loading c:\users\guguer\documents\visual studio 2013\Projects\ConsoleApplication2\B.fs]


B.fs(3,19): error FS0039: The type 'DU' is not defined
> 

There is a workaround, which is to include open Project on B.fs, but it's not obvious. On big projects, and for new users, this is a stumbling block that makes things look broken. 

Similar issues happen with modules instead of namespaces.

### PR DESCRIPTION
**Pull request by dsyme**
This fixes #44, where 

    #load "a1.fs"
    #load "a2.fs"

I've noticed this bug before and it is very irritating when preparing #load lists.

was being a given a different intepretation by fsi.exe and fsc.exe when a1.fs and a2.fs both contribute to the same namespace.  This is because of a mistake in how the "implicitly open own namespace" rule is applied in the presence of the magically inserted FSI_dddd prefixes to all the items defined in each file.  The first file uses prefix FSI_0001 to define namespace N, the second uses FSI_0002 to define namespace N, and the "implicit open own namespace" rule meant that FSI_0002.N was being implicitly opened.

As it happens, there is no need to implicitly open using the prefix path FSI_0002.N, and instead we can implicitly open using just N. This is implemented by filtering the first entry in the impicit open path for the pattern FSI_dddd when running F# Interactive.

### CODE REVIEW

**Comment by ovatsus** [comment]
Great! This bug has been annoying me greatly in FSharp.Data, and force me to add a bunch of redundant open statements everywhere

File: tests/fsharp/core/fsi-reload/run.bat [hunk 1]
```diff
@@ -9,6 +9,23 @@ call %~d0%~p0..\..\..\config.bat
...
 
+  if exist test.ok (del /f /q test.ok)
+  "%FSI%" %fsi_flags%  --maxerrors:1 load1.fsx
+  if NOT EXIST test.ok goto SetError
+
+  if exist test.ok (del /f /q test.ok)
+  "%FSI%" %fsi_flags%  --maxerrors:1 load2.fsx
+  if NOT EXIST test.ok goto SetError
+
+  REM Check we can alo compile, for sanity's sake
```
Comment (ovatsus) [inline tests/fsharp/core/fsi-reload/run.bat]: typo: "also"

File: tests/fsharp/core/fsi-reload/run.bat [hunk 2]
```diff
@@ -9,6 +9,23 @@ call %~d0%~p0..\..\..\config.bat
...
+  "%FSC%" load1.fsx
+  @if ERRORLEVEL 1 goto Error
+
+  REM Check we can alo compile, for sanity's sake
```
Comment (ovatsus) [inline tests/fsharp/core/fsi-reload/run.bat]: typo: "also"

**Comment by dsyme** [comment]
@ovatsus - it would be great if you could try out the fix in that context, and/or identify new test cases to add, 

thanks!

**Comment by isaacabraham** [comment]
Awesome stuff. This might be the thing that makes me try out VS2015 and F#4 ... Will try out in next couple of days.

**Comment by ovatsus** [comment]
Sure, I'll try test it soon

**Comment by latkin** [comment]
Applied with https://github.com/Microsoft/visualfsharp/commit/412e6e675d0abdc225c89db39e1cbe7f93addb78

**Comment by isaacabraham** [comment]
I've finally installed VS2015.   @latkin is there a nightly build of the Visual F# tools to test this out on?

**Comment by latkin** [comment]
Not right now, unfortunately. Our internal repo (used to build the previews and shipping bits) is actually a ways behind the git repo at the moment. It is temporarily frozen while the next VS Preview is being prepared. Sometime next week things should get flowing again.

**Comment by isaacabraham** [comment]
OK, sweet. BTW - slightly off topic but upgraded to CTP 5, on top of which I still had to manually install the F#4 bits to use it - I take it that that's something that's planned to be fixed before the final release?

**Comment by latkin** [comment]
There should be F# bits in the box in CTP 5, though it is expected that they will lag behind the tip of this repo by a some amount (don't recall offhand how far back). Is that what you are seeing, or is there no F# support at all?

**Comment by isaacabraham** [comment]
Well, I installed the CTP and tried to do some simple code testing out F# 4 features - I think I was testing out constructors-as-functions e.g.

    [ "http://www.bbc.co.uk"; ""http://www.tottenhamhotspur.com"] |> List.map System.Uri

But it simply wouldn't compile through FSI, tet F#4 did show up as an items in Extensions and Updates. Only after I manually installed the latest F# bits did it compile. It's possible it was something wrong with my install - I didn't install CTP5 from clean, I installed the V2015 preview and then did CTP5 as an update on top of it - perhaps that was why.

**Comment by ovatsus** [comment]
Tried this with some more complex cases in FSharp.Data after removing all the workarounds, and it worked great!

**Pull request closed without merging**